ethereum/contracts/BasicInboundChannel.sol

```diff
@@ -22,60 +22,39 @@ contract BasicInboundChannel {
+        nonce = 0;
     }
 
     // TODO: add docs
     function submit(
         Message[] calldata _messages,
-        ParachainLightClient.OwnParachainHeadPartial _ownParachainHeadPartial,
-        bytes32[] memory _parachainHeadsProof,
-        BeefyLightClient.BeefyMMRLeafPartial _beefyMMRLeafPartial,
+        ParachainLightClient.OwnParachainHeadPartial
+            calldata _ownParachainHeadPartial,
+        bytes32[] calldata _parachainHeadsProof,
+        ParachainLightClient.BeefyMMRLeafPartial calldata _beefyMMRLeafPartial,
         uint256 _beefyMMRLeafIndex,
         uint256 _beefyMMRLeafCount,
-        bytes32[] memory _beefyMMRLeafProof
+        bytes32[] calldata _beefyMMRLeafProof
     ) public {
-
-        // Require there is enough gas to play all messages
-        require(
-            gasleft() >= _messages.length * MAX_GAS_PER_MESSAGE,
-            "insufficient gas for delivery of all messages"
-        );
-
         // Proof
         // 1. Compute our parachain's message `commitment` by ABI encoding and hashing the `_messages`
-        // TODO
-        // bytes32 commitment = keccak256(abi.encodePacked(_messages));
-
-        // 2. Compute `ownParachainHead` by hashing the data of the `commitment` together with the contents of
-        // `_ownParachainHeadPartial`
-        // TODO
-        // bytes32 ownParachainHead = keccak256(abi.encodePacked(ParachainLightClient.OwnParachainHead(
-        //     _ownParachainHeadPartial.parentHash,
-        //     _ownParachainHeadPartial.number,
-        //     _ownParachainHeadPartial.stateRoot,
-        //     _ownParachainHeadPartial.extrinsicsRoot,
-        //     commitment,
-        // )));
-
-        // 3. Compute `parachainHeadsRoot` by verifying the merkle proof using `ownParachainHead` and
-        // `_parachainHeadsProof`
-        // TODO
-
-        // 4. Compute the `beefyMMRLeaf` using `parachainHeadsRoot` and `_beefyMMRLeafPartial`
-        // TODO
+        bytes32 commitment = keccak256(abi.encode(_messages));
+
+        ParachainLightClient.verifyCommitmentInParachain(
+            commitment,
+            _ownParachainHeadPartial,
+            _parachainHeadsProof,
+            _beefyMMRLeafPartial,
+            _beefyMMRLeafIndex,
+            _beefyMMRLeafCount,
+            _beefyMMRLeafProof
+        );
 
-        // 5. Verify inclusion of the beefy MMR leaf in the beefy MMR root using that `beefyMMRLeaf` as well as
-        // `_beefyMMRLeafIndex`, `_beefyMMRLeafCount` and `_beefyMMRLeafProof`
+        // Require there is enough gas to play all messages
         require(
-            beefyLightClient.verifyBeefyMerkleLeaf(
-                beefyMMRLeaf,
-                _beefyMMRLeafIndex,
-                _beefyMMRLeafCount,
-                _beefyMMRLeafProof
-            ),
-            "Invalid proof"
+            gasleft() >= (_messages.length * MAX_GAS_PER_MESSAGE) + GAS_BUFFER,
+            "insufficient gas for delivery of all messages"
         );
```

Comment:
I did actually move this up because the check is cheap and we could fail quicker that way

Reply:
ah but then we'd need to make sure there is also a buffer to check for doing all the proof checks and everything, so more complex to analyse securely, so i thought just safer to put it at the end and add a small simple buffer